third_party/kmod/eret_hvc_smc_module.c

```diff
@@ -0,0 +1,115 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/*
+ * Copyright (c) 2019 Google LLC
+ *
+ **/
+
+#include <linux/fs.h>
+#include <linux/init.h>
+#include <linux/kernel.h>
+#include <linux/kobject.h>
+#include <linux/module.h>
+#include <linux/printk.h>
+#include <linux/slab.h>
+#include <linux/sysfs.h>
+#include <linux/uaccess.h>
+
+MODULE_LICENSE("GPL");
+MODULE_AUTHOR("Google");
+MODULE_DESCRIPTION("");
+MODULE_VERSION("0.1");
+
+#ifndef __aarch64__
+#  error Unsupported CPU. ARM64 required.
+#endif
+
+// Provides an endpoint on
+//   /sys/kernel/eret_hvc_smc/address
+// to which userspace programs can send their addresses that will be
+// speculatively fetched after speculating over ERET, HVC and SMC instructions.
+// Currently should be accessible only by root, because there is no checking of
+// those addresses.
+
+static ssize_t address_store(struct kobject *kobj, struct kobj_attribute *attr,
+                             const char *buf, size_t length) {
+  size_t address;
+  int *memory, res = kstrtoul(buf, 0, &address);
+  char *pointer;
+
+  if (res != 0) {
+    // Incorrectly formatted address was provided.
+    pr_err("kstrtol failed with %d\n", res);
+    return length;
+  }
+
+  pointer = (char *) address;
+
+  __uaccess_enable(ARM64_ALT_PAN_NOT_UAO);
+  memory = kmalloc(sizeof(int), GFP_KERNEL);
+  memory[0] = (int) length;
+
+  // Core functionality.
+  asm volatile(
+      // 1000 repetitions to confuse the PHT enough.
+      ".rept 1000\n"
+      // Flush "memory[0]" from cache and sync.
+      "dc civac, %0\n"
+      "dsb sy\n"
+      // Slowly load "memory[0]" from main memory and speculate forward in the
+      // meantime.
+      "ldr w1, [%0]\n"
+      "cmn w1, #0x1\n"
+      "bne 1f\n"
+      // Dead code begins - the four following instructions are executed only
+      // speculatively.
+      "eret\n"
+      "hvc #0\n"
+      "smc #0\n"
```

Comment:
Have we demonstrated we can successfully have the kernel speculate past a _real_, _taken_ `ERET? If so, that would make a much more compelling example.

Likewise, we should be showing speculation past a real, taken `svc` in usermode.

As it is, this feels much more like Spectre V1.

Reply:
If I call a real ERET in the kernel module, it will be very complex and very fragile, because I would have to copy-paste all the code that is executed after a sysfs write handler in the kernel. Otherwise it would crash immediately after the ERET (page mapping issues, etc.) or during the next SYSFS write (inconsistent state of the SYSFS). And the maintainability will be also very poor. That is why I quite strongly prefer speculative execution of these kernel instructions. In the case of SVC it is easy to speculate past its real execution, because in userspace it is possible to crash without too much harm. Otherwise it is true that is uses Spectre v1 in the kernel code.

Reply:
Or you know about a scenario where you can call ERET in a kernel module in a simple, robust and maintainable manner?

Reply:
Okay, definitely agree that putting code we control after a will-be-taken `ERET` on Linux without crashing the system is not straightforward.

But my sense is that it would be really valuable to demonstrate the processor actually speculating past an instruction that actually has effects. It provides proof, for example, that the Linux kernel is right to [follow every `ERET` with a speculation barrier](https://patchwork.kernel.org/patch/10700361/).

Maybe we add this to our wishlist of things we'll do when we're using a custom kernel in virtualization.

Reply:
OK.